src/Framework/ReuseableStringBuilder.cs

```diff
@@ -0,0 +1,346 @@
+            /// This constant has to exactly 2^n (power of 2) where n = 4 ... 32
+            /// </summary>
+            /// <remarks>
+            /// This constant might looks huge, but rather that lowering this constant,
+            ///   we shall focus on eliminating of code which requires to create such huge strings.
+            /// </remarks>
+            private const int MaxBuilderSizeBytes = 2 * 1024 * 1024; // ~1M chars
+            private const int MaxBuilderSizeCapacity = MaxBuilderSizeBytes / 2;
+
+            private static readonly IReadOnlyList<int> s_capacityBrackets;
+
+            static ReuseableStringBuilderFactory()
+            {
+                var brackets = new List<int>();
+
+                int bytes = 0x200; // Minimal capacity is 256 (512 bytes) as this was, according to captured traces, mean required capacity
+                while (bytes <= MaxBuilderSizeBytes)
+                {
+                    // Allocation of arrays is optimized in byte[bytes] => bytes = 2^n.
+                    // StringBuilder allocates chars[capacity] and each char is 2 bytes so lets have capacity brackets computed as `bytes/2` 
+                    brackets.Add(bytes/2); 
+                    bytes <<= 1;
+                }
+                Debug.Assert((bytes >> 1) == MaxBuilderSizeBytes, "MaxBuilderSizeBytes has to be 2^n (power of 2)");
+
+                s_capacityBrackets = brackets;
+            }
+
+            /// <summary>
+            /// The shared builder.
+            /// </summary>
+            private static StringBuilder s_sharedBuilder;
+
+#if DEBUG && ASSERT_BALANCE
+            /// <summary>
+            /// Balance between calling Get and Release.
+            /// Shall be always 0 as Get and 1 at Release.
+            /// </summary>
+            private static int s_getVsReleaseBalance;
+#endif
+
+            /// <summary>
+            /// Obtains a string builder which may or may not already
+            /// have been used. 
+            /// Never returns null.
+            /// </summary>
+            internal static StringBuilder Get(int capacity)
+            {
+#if DEBUG && ASSERT_BALANCE
+                int balance = Interlocked.Increment(ref s_getVsReleaseBalance);
+                Debug.Assert(balance == 1, "Unbalanced Get vs Release. Either forgotten Release or used from multiple threads concurrently.");
+#endif
+
+                var returned = Interlocked.Exchange(ref s_sharedBuilder, null);
+
+                if (returned == null)
+                {
+                    // Currently loaned out so return a new one with capacity in given bracket.
+                    // If user wants bigger capacity that maximum capacity respect it.
+                    returned = new StringBuilder(SelectBracketedCapacity(capacity));
+#if DEBUG
+                    MSBuildEventSource.Log.ReusableStringBuilderFactoryStart(hash: returned.GetHashCode(), newCapacity:capacity, oldCapacity:0, type:"miss");
+#endif
```

Comment:
It's a bit odd to event only on DEBUG. Is it super costly to do it all the time? Should we just check for is-event-enabled?

Reply:
In Roslyn build it was ~600K events in default console verbosity and ~1.2M in diag file log verbosity.
Since the limit in perfview is, IIRC, 20M this could harm generation of production or some customer ETW trace collections.
I considered these ETW usable only for development from time to time to verify that out presumption about ReusableStringBuilderFactory usage are still valid.

Reply:
Sounds reasonable. Can you comment that somewhere, maybe on the events themselves?